src/http-api/resources/files.js

```diff
@@ -0,0 +1,50 @@
+'use strict'
+
+const bs58 = require('bs58')
+const debug = require('debug')
+const log = debug('http-api:files')
+log.error = debug('http-api:files:error')
+
+exports = module.exports
+
+// common pre request handler that parses the args and returns `key` which is assigned to `request.pre.args`
+exports.parseKey = (request, reply) => {
+  if (!request.query.arg) {
+    return reply("Argument 'key' is required").code(400).takeover()
+  }
+
+  try {
+    return reply({
+      key: new Buffer(bs58.decode(request.query.arg))
+    })
+  } catch (err) {
+    log.error(err)
+    return reply({
+      Message: 'invalid ipfs ref path',
+      Code: 0
+    }).code(500).takeover()
+  }
+}
+
+exports.cat = {
+  // uses common parseKey method that returns a `key`
+  parseArgs: exports.parseKey,
+
+  // main route handler which is called after the above `parseArgs`, but only if the args were valid
+  handler: (request, reply) => {
+    const key = request.pre.args.key
+
+    request.server.app.ipfs.files.cat(key, (err, stream) => {
+      if (err) {
+        log.error(err)
+        return reply({
+          Message: 'Failed to cat file: ' + err,
+          Code: 0
+        }).code(500)
+      }
+      stream.on('data', (data) => {
+        return reply(data.stream)
```

Comment:
It *probably* will be okay, but I think @dignifiedquire's point was about approaching the mechanism defensively: just because today we can reason out that this should never happen doesn't mean future changes by future authors will know to hold this requirement. "It's a stream after all, and you can have many `data` events on streams.."